File: Python/BinarySearchTree/bs_tree.py
#Author: Stephen Pryor
#Date: September 25, 12

#-------------------------------------------START - subtree
class subtree:
  def __init__(self, value):
    self.val = value
    self.left = None
    self.right = None
#-------------------------------------------END - subtree

#-------------------------------------------START - bs_tree
class bs_tree:
  #constructor
  def __init__(self):
    self.root = None
    self.n = 0
  
  #START - Empty Function
  def empty(self):
    return self.root == None
  #END - Empty Function
  
  #START - Size Function
  def size(self):
    return self.n
  #END - Size Function
  
  #START - Insertion Function
  def insert(self, value):
    if self.empty():
      self.root = subtree(value)
      self.n = self.n + 1
    else:
      self._insert(value, self.root)
  
  #recursive insert      
  def _insert(self, value, leaf):
    if value > leaf.val:
      if leaf.right == None:
        leaf.right = subtree(value)
        self.n = self.n + 1
      else:
        self._insert(value, leaf.right)
    elif value < leaf.val:
      if leaf.left == None:
        leaf.left = subtree(value)
        self.n = self.n + 1
      else:
        self._insert(value, leaf.left)
  #END - Insertion Function
  
  #START - Search Function
  def search(self, value):
    return self._search(value, self.root)
  
  #recursive search
  def _search(self, value, leaf):
    if leaf == None:
      return False
    elif value == leaf.val:
      return True
    elif value > leaf.val:
      return self._search(value, leaf.right)
    else:
      return self._search(value, leaf.left)
  #END - Search Function
  
  #START - Deletion Function
  def remove(self, value):
    if not self.empty():
      self.root = self._remove(value, self.root)
  
  #recursive remove
  def _remove(self, value, leaf):
    if leaf != None:
      if value == leaf.val:
        if leaf.left == None and leaf.right == None:
          leaf = None
        elif leaf.left != None and leaf.right == None:
          leaf = leaf.left
        else:
          if leaf.right.left == None:
            leaf.right.left = leaf.left
            leaf = leaf.right
          else:
            prevLeaf = leaf.right 
            nextLeaf = leaf.right.left 
            while nextLeaf.left != None:
               prevLeaf = nextLeaf
               nextLeaf = nextLeaf.left
            leaf.val = nextLeaf.val
            prevLeaf.left = nextLeaf.right
        self.n = self.n - 1
      elif value > leaf.val:
        leaf.right = self._remove(value, leaf.right)
      elif value < leaf.val:
        leaf.left = self._remove(value, leaf.left)
    return leaf
  #END - Deletion Function

  #START - Inorder Traversal Function
  def print_inorder(self):
    self._print_inorder(self.root)
  
  #recursive print inorder 
  def _print_inorder(self, leaf):
    if leaf != None:
      self._print_inorder(leaf.left)
      print leaf.val
      self._print_inorder(leaf.right)
  #END - Inorder Traversal Function
#-------------------------------------------END - bs_tree

File: C++/BinarySearchTree/bs_tree.cpp
//Author: Stephen Pryor
//This is a quick Binary Search Tree data structure.
//I have not fully tested it but I believe it is correct.
//There is likely room for improvement in the following code. 

#include <iostream>
#include "subtree.cpp"

using namespace std;

template<class T>
class bs_tree {
  private:
    subtree<T> * root;
    int n;
    
    void _insert(T value, subtree<T> * &leaf);
    void _remove(T value, subtree<T> * &leaf);
    bool _search(T value, subtree<T> * leaf);
    void _print_inorder(subtree<T> * leaf);
    void destroy(subtree<T> * &leaf);
  public:
    bs_tree();
    ~bs_tree();
    bool empty();
    int size();
    void insert(T value);
    void remove(T value);
    bool search(T value);
    void print_inorder();
};

template <class T>
bs_tree<T>::bs_tree() {
  root = 0;
  n = 0;
}

template <class T>
bs_tree<T>::~bs_tree() {
  destroy(root);
}

template <class T>
void bs_tree<T>::destroy(subtree<T> * &leaf) {
  if(leaf != 0)
  {
    destroy(leaf.left);
    destroy(leaf.right); 
  }
  delete leaf;
  leaf = 0;
}

template <class T>
bool bs_tree<T>::empty() {
  return root == 0;
}

template <class T>
int bs_tree<T>::size() {
  return n;
}

template <class T>
void bs_tree<T>::insert(T value) {
  if(empty()) {
    root = new subtree<T>(value);
    ++n;
  } else {
    _insert(value, root);
  }
}

template <class T>
void bs_tree<T>::_insert(T value, subtree<T> * &leaf) {
  if(leaf == 0)
  {
    leaf = new subtree<T>(value);
    ++n;
  } else if(leaf->value > value) {
    _insert(value, leaf->left);
  } else if(leaf->value < value) {
    _insert(value, leaf->right);
  }
}

template <class T>
void bs_tree<T>::remove(T value) {
  _remove(value, root);
}

template <class T>
void bs_tree<T>::_remove(T value, subtree<T> * &leaf) {
  if(leaf != 0) {
    if(leaf->value == value) {
      if(leaf->left == 0 && leaf->right == 0)
      {
        delete leaf;
        leaf = 0;
      } else if(leaf->left != 0 && leaf->right == 0) {
        subtree<T> * tmp = leaf;
        leaf = leaf->left;
        delete tmp;
        tmp = 0;
      } else {
        if(leaf->right->left == 0)
        {
          subtree<T> * tmp = leaf;
          leaf->right->left = leaf->left;
          leaf = leaf->right;
          delete tmp;
          tmp = 0;
        } else {
          subtree<T> * prevLeaf = leaf->right;
          subtree<T> * nextLeaf = leaf->right->left;
          while(nextLeaf->left != 0) {
            prevLeaf = nextLeaf;
            nextLeaf = nextLeaf->left;
          }
          leaf->value = nextLeaf->value;
          prevLeaf->left = nextLeaf->right;
          delete nextLeaf;
          nextLeaf = 0;
        }
      }
      --n;
    } else if(leaf->value < value) {
      _remove(value, leaf->right);
    } else if(leaf->value > value) {
      _remove(value, leaf->left);
    }
  }
}

template <class T>
bool bs_tree<T>::search(T value) {
  return _search(value, root);
}

template <class T>
bool bs_tree<T>::_search(T value, subtree<T> * leaf) {
  if(leaf == 0) {
    return false;
  } else if(leaf->value == value) {
    return true;
  } else if(leaf->value > value) {
    return _search(value, leaf->left);
  } else {
    return _search(value, leaf->right);
  }
}

template <class T>
void bs_tree<T>::print_inorder() {
  _print_inorder(root);
}

template <class T>
void bs_tree<T>::_print_inorder(subtree<T> * leaf) {
  if(leaf != 0){
    _print_inorder(leaf->left);
    cout << leaf->value << endl;
    _print_inorder(leaf->right);
  }
}
